### ISSUE TITLE
GEP: TLS from Gateway to Backend for ingress (backend TLS termination)

### ISSUE THREAD

**Issue by candita** [kubernetes-sigs/gateway-api#1897]
**What would you like to be added**:
Add a document which specifically addresses the topic of conveying HTTPS from the Gateway dataplane to the backend (backend TLS termination), and intends to satisfy the single use case “As a client implementation of Gateway API, I need to know how to connect to a backend pod that has its own certificate”. TLS configuration can be a nebulous topic, so in order to drive resolution this GEP focuses only on this single piece of functionality. 

Given that the current ingress solution specifies **edge** TLS termination (from the edge to the gateway), and how to handle **passthrough** TLS (from the edge to the backend pod), this proposed ingress solution specifies **backend** TLS termination (from the gateway to the backend pod).  As mentioned, this solution satisfies the use case in which the backend pod has its own certificate and the gateway client needs to know how to connect to the backend pod.

**Why this is needed**:

The document is _very_ tightly scoped because we have tried and failed to address this well-known gap in the API specification. The lack of support for this fundamental concept is holding back Gateway API adoption by users that require a solution to the use case. One of the recurring themes that has held up the prior art has been interest related to service mesh and as such this proposal focuses only on the ingress use case to avoid contention there.  Another reason for the tight scope is that we have been too focused on a generic representation of everything that TLS can do, which covers too much ground to address in a single GEP.

**Comment by dprotaso** [comment]
Note: this GEP is pulling TLS use cases from this earlier proposal - https://github.com/kubernetes-sigs/gateway-api/issues/1282

**Comment by youngnick** [comment]
We actually haven't had much experience yet with finalizing this state, but either one of those seems reasonable to me for now.

**Comment by candita** [comment]
Planning to start conformance testing soon.
/remove-lifecycle stale

**Comment by robscott** [comment]
I'm thinking that we may want to narrow the scope of the title of this GEP to reflect the scope that BackendTLSPolicy launched with in v1.0. That policy is focused on verifying the certificates served by the backend. The other part of TLS from Gateway to Backend would be configuring a client cert that a Gateway should use to connect to a Backend. In my opinion, that's probably best accomplished with a separate GEP. If everyone's OK with that, I'll adjust the title of this GEP to reflect that narrower scope and create a new GEP issue to track configuring how the Gateway connects to backends.

**Comment by kfox1111** [comment]
Was a new GEP for client certs ever filed? I'd be interested in it. In particular, if SPIFFE/SPIRE could be used as the client/remote certs.

**Comment by albionb96** [comment]
> Was a new GEP for client certs ever filed? I'd be interested in it. In particular, if SPIFFE/SPIRE could be used as the client/remote certs.

I would be also very interested to know what is the status of this?

**Comment by candita** [comment]
@albionb96 @kfox1111 the new GEP for client certs is https://gateway-api.sigs.k8s.io/geps/gep-91/  Not sure if this is what you're looking for, but 

> This GEP proposes a way to validate the TLS certificate presented by the frontend client to the server (Gateway Listener in this case) during a [TLS Handshake Protocol](https://www.rfc-editor.org/rfc/rfc5246#section-7.4).

**Comment by albionb96** [comment]
@candita thank you for your response. I did a quick research and came up with some results (please correct me if I am wrong):
![image](https://github.com/kubernetes-sigs/gateway-api/assets/49485798/9a6aca39-1d8c-4934-b2ea-b125f78db064)

Number 1 (Client-Cert): will be addressed with this GEP: https://gateway-api.sigs.k8s.io/geps/gep-91/
Number 2 (Server-Cert): Is already implemented and well-known where the Gateway API sends its cert to the Client for verification.

Number 3 (Client-Cert): will be addressed with this GEP: https://gateway-api.sigs.k8s.io/geps/gep-1897/?h=1897+gep
Number 4 (Server-Cert): Is addressed in the experimental channel with backendTLSPolicy: https://gateway-api.sigs.k8s.io/api-types/backendtlspolicy/

First question: If all of these GEPs are implemented would we then be able to achieve the "re-encrpyt" mode, with mutual authentication on both sides backend and frontend?
e.g.
FrontendClient<-mTLS->GatewayAPI<-mTLS->BackendService

or did I get it totally wrong and mutual authentication on both ends, (backend and frontend), or only one end and also re-encrypt is not planed at all?

Second question: even if I got everything correct and we will have re-encrypt with mutual authentication on both ends, how will be the identity of the client forwarded to the backend, because for me the frontendValdiation looks like a check up on the caCertificateRefs and that's it, the new connection from gateway API to backend won't know anything about the client or the requestor, right?

For example when using an envoy proxy it is possible to forward the identity of the requestor (frontendclient), using the XFCC-Header, is something like this possible in gatewayAPI also, or is this planned to happen somehow in the future?

Thank you very much in advance!

**Comment by youngnick** [comment]
> First question: If all of these GEPs are implemented would we then be able to achieve the "re-encrpyt" mode, with mutual authentication on both sides backend and frontend? e.g. FrontendClient<-mTLS->GatewayAPI<-mTLS->BackendService

Yes. Although because of the crossover with Service Mesh's use of "mTLS", we've tended to be more specific about our language. But assuming you're meaning "mTLS" as "a TLS transaction where the client and server both validate each other's certificates" (and not "automatically provisioned and rotated invisibly by infrastructure TLS between a services", which is what Service Mesh contexts usually mean by "mTLS"), then _yes_, those GEPs should give that.

Again, we have used the term "re-encrypt" in the past, but have found it can mean different things to different people, so we try to be more specific.

> Second question: even if I got everything correct and we will have re-encrypt with mutual authentication on both ends, how will be the identity of the client forwarded to the backend, because for me the frontendValdiation looks like a check up on the caCertificateRefs and that's it, the new connection from gateway API to backend won't know anything about the client or the requestor, right?

As of now, there are no plans to do identity forwarding. But it's certainly something we would consider if someone raised the request.

**Comment by youngnick** [comment]
/remove-lifecycle stale

Definitely not stale, still being worked on.

**Comment by youngnick** [comment]
/remove-lifecycle rotten

Still in progress.

**Comment by candita** [comment]
Also https://github.com/kubernetes-sigs/gateway-api/issues/3516

**Comment by shaneutt** [comment]
This feature has been accepted for the `v1.4.0` release. Please see [this announcement](https://github.com/kubernetes-sigs/gateway-api/discussions/3824) for more details. As I understand it this one is most of the way there, mostly in need of a completion of conformance tests and then some conformance reports to be submitted from implementations. If you have any questions, concerns, or are in need of support please reach out to the [maintainers](https://github.com/kubernetes-sigs/gateway-api/blob/e7c711b0a91273727288942faa10573164e6981b/OWNERS_ALIASES#L13) so we can assist you!

@candita on the [standard features submissions](https://github.com/kubernetes-sigs/gateway-api/discussions/3758) you marked down "yes" for reviewers, but I don't see the actual reviewers. Please make sure you have commitments from at least one (and ideally two) reviewers to help you move this forward, and ping them here.

**Comment by howardjohn** [comment]
I can act as a reviewer

**Comment by candita** [comment]
@snorwin @howardjohn @Miciah thanks for your help with reviews.

**Comment by candita** [comment]
Also, thanks @kl52752 for help with reviews.

**Comment by shaneutt** [comment]
Just leaving a note here to follow up on: https://github.com/kubernetes-sigs/gateway-api/pull/3563

**Comment by shaneutt** [comment]
This issue is targeting a new `Standard` API in the the [v1.4.0 release-cycle](https://github.com/kubernetes-sigs/gateway-api/discussions/3969), so this is just a reminder that we're looking to do code-freeze on August 26th, which is **two weeks from now**.

@candita just a quick check-in can you please enumerate the remaining work that needs to be done, if not already covered by the sub-tasks of this issue, and a brief assessment of where those tasks are in any case? I know we're missing some tests, I want to make sure we're gathering anything else (ideally as subtasks) and coming up with a plan forward.

**Comment by shaneutt** [comment]
Based on the request from extension [here](https://github.com/kubernetes-sigs/gateway-api/pull/4009#issuecomment-3223036067) from @kl52752:

> @shaneutt I would like to request the extension for this GEP and include it in v1.4.0 release in experimental channel.
> We need to reach out agreement for naming convention and I think that we are very close and need 24h to finalize this.

Extension granted until September 3rd for this and all sub-tasks to finish up. @kl52752 and @snorwin (who have the remaining subtasks) please stay in touch and ping @kubernetes-sigs/gateway-api-admins  if you get blocked or need support to keep things moving forward. After Sept 3 we'll start locking in v1.4.0 for code freeze; if more time is needed at that point, we'll switch things over to v1.5.0 (which planning for that will be starting in a few weeks here).

**Comment by shaneutt** [comment]
> Based on the request from extension [here](https://github.com/kubernetes-sigs/gateway-api/pull/4009#issuecomment-3223036067) from [@kl52752](https://github.com/kl52752):
> 
> > [@shaneutt](https://github.com/shaneutt) I would like to request the extension for this GEP and include it in v1.4.0 release in experimental channel.
> > We need to reach out agreement for naming convention and I think that we are very close and need 24h to finalize this.
> 
> Extension granted until September 3rd for this and all sub-tasks to finish up. [@kl52752](https://github.com/kl52752) and [@snorwin](https://github.com/snorwin) (who have the remaining subtasks) please stay in touch and ping [@kubernetes-sigs/gateway-api-admins](https://github.com/orgs/kubernetes-sigs/teams/gateway-api-admins) if you get blocked or need support to keep things moving forward. After Sept 3 we'll start locking in v1.4.0 for code freeze; if more time is needed at that point, we'll switch things over to v1.5.0 (which planning for that will be starting in a few weeks here).

We're all set, this made it in time and is releasing in `v1.4.0` as standard.
